cpp/src/arrow/compute/kernels/scalar_temporal.cc

```diff
@@ -321,6 +322,82 @@ struct Nanosecond {
   }
 };
 
+// ----------------------------------------------------------------------
+// Convert timestamps to a string representation with an arbitrary format
+
+template <typename Duration>
+inline std::string get_timestamp(int64_t arg, const StrftimeOptions* options) {
+  auto zt = arrow_vendored::date::zoned_time<Duration>{options->tz,
+                                                       sys_time<Duration>(Duration{arg})};
+  return arrow_vendored::date::format(options->format, zt);
+}
+
+// If source timestamp is timezone naive we do not print timezone info
+inline StrftimeOptions get_options(KernelContext* ctx, const bool timezone_known) {
+  const StrftimeOptions options = StrftimeState::Get(ctx);
+  if (timezone_known) {
+    return options;
+  }
+
+  std::string new_format = std::move(options.format);
+  for (std::string str : {"%z", "%Z"}) {
+    size_t pos = std::string::npos;
+    while ((pos = new_format.find(str)) != std::string::npos) {
+      new_format.erase(pos, str.length());
+    }
+  }
+  return StrftimeOptions(new_format, options.timezone);
+}
+
+template <typename Duration>
+struct Strftime {
+  static Status Call(KernelContext* ctx, const Scalar& in, Scalar* out) {
+    const auto& timezone = GetInputTimezone(in);
+    const StrftimeOptions options = get_options(ctx, !timezone.empty());
+    if (timezone.empty() && options.timezone != "UTC") {
+      return Status::Invalid("Timezone naive timestamp can only be printed in UTC. Got: ",
```

Comment:
Why allow UTC here (and not other timezones, or disallow any timezone including UTC) ?

Reply:
My thinking was to have an option to print timezone naive timestamps at least as UTC. I'll disallow printing of timezone naive timestamps.

Reply:
Removed `options.timezone` so this logic was removed.